modules/ephemeris/src/main/scala/gem/horizons/HorizonsClient.scala

```diff
@@ -30,11 +31,14 @@ import scala.concurrent.duration._
   */
 object HorizonsClient {
 
-  /** Horizons permits only one request at a time from a given host, so all
-    * requests should go through this client.
+  /** A stream that emits a single client that will be shut down automatically. An internal global
+    * mutex ensures that only one such stream exists at a time.
     */
-  val client: IO[Client[IO]] =
-    Http1Client[IO](BlazeClientConfig.defaultConfig.copy(responseHeaderTimeout = 20.seconds))
+  val client: Stream[IO, Client[IO]] = {
+    val mutex  = Semaphore[IO](1L)(implicitly, scala.concurrent.ExecutionContext.global).unsafeRunSync // note
+    val client = Http1Client.stream[IO](BlazeClientConfig.defaultConfig.copy(responseHeaderTimeout = 20.seconds))
+    Stream.bracket(mutex.decrement)(_ => client, _ =>  mutex.increment)
+  }
```

Comment:
This is very clever.  I don't think I would have come up with that solution.  It streams clients but only one at a time due to the `mutex`, correct?

Reply:
Correct. We take the mutex before running the `client` stream, then when the stream terminates the client is cleaned up (internally it also uses `bracket` to guarantee this) and then the lock is released. If the lock is already taken we don't block the thread. We're suspended and later rescheduled on the mutex's `ExecutionContext`.